integrations/redis.go

```diff
@@ -0,0 +1,125 @@
+package integrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-redis/redis/v7"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+const (
+	fail    = "fail"
+	success = "success"
+)
+
+type redisContextKey int
+
+const (
+	redisSpanName redisContextKey = iota
+	redisCommandTimer
+)
+
+func setRedisSpanName(ctx context.Context, name string) context.Context {
+	return context.WithValue(ctx, redisSpanName, name)
+}
+
+func getRedisSpanName(ctx context.Context) (name string) {
+	if n, ok := ctx.Value(redisSpanName).(string); ok {
+		name = n
+		return
+	}
+	return
+}
+
+// RedisBaseplateHook is a redis.Hook for wrapping Redis commands and pipelines
+// in Client Spans and metrics.
+type RedisBaseplateHook struct {
+	ClientName string
+	Metrics    metricsbp.Statsd
+}
+
+var _ redis.Hook = RedisBaseplateHook{}
+
+// BeforeProcess starts a client Span before processing a Redis command and
+// starts a timer to record how long the command took.
+func (h RedisBaseplateHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, cmd.Name()), nil
+}
+
+// AfterProcess ends the client Span started by BeforeProcess, publishes the
+// time the Redis command took to complete, and a metric indicating whether the
+// command was a "success" or "fail"
+func (h RedisBaseplateHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	return h.endChildSpan(ctx, cmd.Err())
+}
+
+// BeforeProcessPipeline starts a client span before processing a Redis pipeline
+// and starts a timer to record how long the pipeline took.
+func (h RedisBaseplateHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, "pipeline"), nil
+}
+
+// AfterProcessPipeline ends the client span started by BeforeProcessPipeline,
+// publishes the time the Redis pipeline took to complete, and a metric
+// indicating whether the pipeline was a "success" or "fail"
+func (h RedisBaseplateHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var err error
+	for _, cmd := range cmds {
+		err = cmd.Err()
+		if err != nil {
+			break
+		}
+	}
+	return h.endChildSpan(ctx, err)
+}
+
+func (h RedisBaseplateHook) startChildSpan(ctx context.Context, cmdName string) context.Context {
+	span := tracing.GetServerSpan(ctx)
+	if span == nil {
```

Comment:
when there's no server span, do we still want the timer metrics?

Reply:
Thanks to #41 we don't even have to do this in here anymore, we can just rely on the metrics from the baseplate span hook. :)